src/test/poll/test_poll.rs

```diff
@@ -0,0 +1,223 @@
+ * See LICENSE for licensing information
+ */
+
+use test_utils::set;
+use test_utils::TestEnvironment as TestEnv;
+
+fn fd_write(fd: i32) -> i32 {
+    return unsafe { libc::write(fd, b"test".as_ptr() as *const libc::c_void, 4) as i32 };
+}
+
+fn fd_read_cmp(fd: i32) -> Result<(), String> {
+    let mut buf = [0_u8; 4];
+    nix::unistd::read(fd, &mut buf).map_err(|e| e.to_string())?;
+    println!("read from buf: {:?}", buf);
+    if &buf != b"test" {
+        return Err("error: did not read 'test' from pipe.".to_string());
+    } else {
+        return Ok(());
+    }
+}
+
+fn test_pipe() -> Result<(), String> {
+    /* Create a set of pipefds
+    pfd[0] == read, pfd[1] == write */
+    let mut pfds = [0_i32; 2];
+    if unsafe { libc::pipe(pfds.as_mut_ptr() as *mut libc::c_int) < 0 } {
+        return Err("error: pipe could not be created!".to_string());
+    }
+
+    /* poll will check when pipe has info to read */
+    let mut read_poll = libc::pollfd {
+        fd: pfds[0],
+        events: libc::POLLIN,
+        revents: 0,
+    };
+
+    /* First make sure there's nothing there */
+    let mut ready = unsafe { libc::poll(&mut read_poll as *mut libc::pollfd, 1, 100) };
+    if ready < 0 {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err("error: poll failed".to_string());
+    } else if ready > 0 {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err(format!(
+            "error: pipe marked readable. revents={}",
+            read_poll.revents
+        ));
+    }
+
+    /* Now put information in pipe to be read */
+    if fd_write(pfds[1]) < 0 {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err(format!("error: could not write to pipe"));
+    }
+
+    /* Check again, should be something to read */
+    read_poll.fd = pfds[0];
+    read_poll.events = libc::POLLIN;
+    read_poll.revents = 0;
+    ready = unsafe { libc::poll(&mut read_poll as *mut libc::pollfd, 1, 100) };
+    if ready != 1 {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err(format!("error: poll returned {} instead of 1", ready));
+    }
+
+    if read_poll.revents & libc::POLLIN == 0 {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err(format!(
+            "error: read_poll has wrong revents: {}",
+            read_poll.revents
+        ));
+    }
+
+    /* Make sure we got what expected back */
+    if let Err(e) = fd_read_cmp(pfds[0]) {
+        unsafe {
+            libc::close(pfds[0]);
+            libc::close(pfds[1]);
+        }
+        return Err(e);
+    }
+
+    /* success! */
+    unsafe {
+        libc::close(pfds[0]);
+        libc::close(pfds[1]);
+    }
+    return Ok(());
+}
+
+fn test_creat() -> Result<(), String> {
```

Comment:
This function can use the `test_utils::run_and_close_fds` wrapper as well to not require explicit `libc::close()` calls.